Add an option to keep native debug symbols

omajid opened the pull request:
When packaging .NET for Linux distributions, the package builders generally use a different workflow for shipping symbols to users:

1. The package maintainer builds code with the debug flags (such as `-g`) to generate full native debug info and symbols.

2. Nothing is stripped from build by the package maintainer.

3. The build system (`rpmbuild`, `debuild`) removes the debug info (or debug symbols) from the code and creates separate `-debuginfo` or `-debug` packages that contain just the debug symbols.

4. These debug packages are then distributed along with the normal packages using the normal Linux distribution mechanisms. This letsusers install the exact set of debug symbols matching their other package.

To support this workflow in dotnet/runtime, we need to add optional support for not stripping debug symbols. I used it has follows:

    CFLAGS=-g CXXFLAGS=-g ./build.sh --keepsymbols true

After this build, the built binaries include all debug symbols.

I can then rely on the distro package build system to identify, strip, package and ship the debug info/symbols separately.

See https://github.com/dotnet/runtime/issues/3781 and https://github.com/dotnet/source-build/issues/267 for more details on the background and motivation.

For some related fixes, see:

- https://github.com/dotnet/coreclr/pull/3445
- https://github.com/dotnet/corefx/pull/24979

omajid commented:
This is a bit of a regression from the equivalent support in dotnet/coreclr and dotnet/corefx

CC'ing some folks who were part of the conversation last time around: @dagood @dleeapho @eerhardt @jkotas @tmds @weshaggard

omajid commented:
Anyone willing to review this?

dagood commented:
@jkoritzinsky maybe you'd like to review this change--cmake conditions and new build script arg.

dagood commented on eng/build.sh:
```diff
@@ -74,6 +74,7 @@ usage()
   echo "  --gcc                      Optional argument to build using gcc in PATH (default)."
   echo "  --gccx.y                   Optional argument to build using gcc version x.y."
   echo "  --portablebuild            Optional argument: set to false to force a non-portable build."
+  echo "  --keepsymbols              Optional argument: set to true to keep symbols/debuginfo in generated binaries."
```
The help message doesn't mention it takes an argument, but it seems to take true/false and reject nothingness. 😕 But it looks like this is also the case with other settings, so this PR appears to be consistent to me.

omajid replied:
Should I take a stab at fixing up the other settings too?

dagood commented on eng/build.sh:
```diff
@@ -144,6 +145,7 @@ cmakeargs=''
 extraargs=''
 crossBuild=0
 portableBuild=1
+keepDebugInfo=0
```
Seems unused? (Maybe from an initial implementation more like `portableBuild` but it ended up being simpler?)

omajid replied:
Thanks, removed now.

dagood reviewed (approved):
Seems very reasonable. One suggestion and one comment about the context--I'm not particularly familiar with this area.

tmds reviewed (approved):
LGTM, thanks!

omajid commented:
@janvorli hey, can you help me find some reviewers for this PR? Thanks in advance!

janvorli commented on eng/build.sh:
```diff
@@ -377,6 +378,18 @@ while [[ $# > 0 ]]; do
+      fi
+      passedKeepSymbols="$(echo "$2" | awk '{print tolower($0)}')"
+      if [ "$passedKeepSymbols" = true ]; then
+        cmakeargs="${cmakeargs} -cmakeargs -DCLR_KEEP_SYMBOLS=true"
```
I would prefer creating a msbuild property for this instead of passing it in the cmakeargs. We do that e.g. for the `-portablebuild` option, so I'd like to use unified way for all of them. You would then add this option to `eng/build-commons.sh`, which is used to parse the common options for coreclr, libraries and installer .sh scripts and add the cmake arg there. You'll need to update src/coreclr/runtime.proj, src/installer/corehost/corehost.proj and src/libraries/Native/build-native.proj to convert the msbuild property to the -keepsymbols option passed to the respective .sh scripts.

omajid replied:
Does the updated patch match what you had in mind?

janvorli commented on eng/build.sh:
```diff
@@ -377,6 +378,18 @@ while [[ $# > 0 ]]; do
+      fi
+      passedKeepSymbols="$(echo "$2" | awk '{print tolower($0)}')"
+      if [ "$passedKeepSymbols" = true ]; then
+        cmakeargs="${cmakeargs} -cmakeargs -DCLR_KEEP_SYMBOLS=true"
```
A nit - based on various cmake args settings in coreclr, I'd prefer this to be named CLR_CMAKE_KEEP_SYMBOLS to keep things unified.

omajid replied:
Done!

janvorli reviewed (changes requested):
I would like to ask you to change few things.

omajid committed:
Add an option to keep native debug symbols

When packaging .NET for Linux distributions, the package builders
generally use a different workflow for shipping symbols to users:

1. The package maintainer builds code with the debug flags (such as
   `-g`) to generate full native debug info and symbols.

2. Nothing is stripped from build by the package maintainer.

3. The build system (`rpmbuild`, `debuild`) removes the debug
   info (or debug symbols) from the code and creates separate
   `-debuginfo` or `-debug` packages that contain just the debug
   symbols.

4. These debug packages are then distributed along with the normal
   packages using the normal Linux distribution mechanisms. This lets
   users install the exact set of debug symbols matching their other
   package.

To support this workflow in dotnet/runtime, we need to add optional
support for not stripping debug symbols. I used it has follows:

    CFLAGS=-g CXXFLAGS=-g ./build.sh --keepnativesymbols true

After this build, the built binaries include all debug symbols.

I can then rely on the distro package build system to identify, strip,
package and ship the debug info/symbols separately.

See https://github.com/dotnet/runtime/issues/3781 and
https://github.com/dotnet/source-build/issues/267 for more details on
the background and motivation.

For some related fixes, see:

- https://github.com/dotnet/coreclr/pull/3445
- https://github.com/dotnet/corefx/pull/24979

janvorli reviewed (approved):
LGTM, thank you!

omajid commented:
@janvorli Does the "keepSymbols" name appear clear enough to you? Should I look at disambiguating it to "keepNativeSymbols" or "keepUnmanagedSymbols" or something similar?

janvorli commented:
Hmm, `keepNativeSymbols` sounds better to me.

omajid commented:
The CI failures look unrelated to my changes (json tests and mono build issues) but I have rebased this branch onto master to see if that fixes the CI failures.

omajid commented:
@janvorli Does this still look okay?

janvorli reviewed (approved):
LGTM with the latest updates.

janvorli commented:
I'll still re-run the runtime tests to see the `runtime (Installer Build and Test coreclr Linux_arm Debug)` not to timeout.
